shared/src/TwinCollectionArray.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// Represents a property array in <see cref="TwinCollection"/>
+    /// </summary>
+    public class TwinCollectionArray : JArray
+    {
+        private const string MetadataName = "$metadata";
+        private const string LastUpdatedName = "$lastUpdated";
+        private const string LastUpdatedVersionName = "$lastUpdatedVersion";
+
+        private readonly JObject _metadata;
+
+        internal TwinCollectionArray(JArray jArray, JObject metadata)
+            : base(jArray)
+        {
+            _metadata = metadata;
```

Comment:
null check?

Reply:
metadata is a JObject, which will default to null anyway. Is there an advantage to adding an explicit null check here?

Reply:
There is no guarding against NRE later in this file.

So either it can be null, in which case we we should conditionally derereference it, or we should check for null and throw.

Reply:
I see, I can throw an ArgumentNullException during initialization, sure. The use-case does require the metadata to be available (and this is an internal constructor), so makes sense.